src/v/cloud_storage/topic_manifest.cc

```diff
@@ -22,12 +23,175 @@
+            } else if ("compaction_strategy" == _key) {
+                _topic_config.properties.compaction_strategy
+                  = boost::lexical_cast<model::compaction_strategy>(
+                    std::string(str, length));
+            } else if ("timestamp_type" == _key) {
+                _topic_config.properties.timestamp_type
+                  = boost::lexical_cast<model::timestamp_type>(
+                    std::string(str, length));
+            } else {
+                return false;
+            }
+            if (_required_fields.contains(_key)) {
+                _seen_required_fields.insert(_key);
+            }
+            _state = all_required_complete() ? expectKeyOrObjectEnd : expectKey;
+            return true;
+        default:
+            return false;
+        }
+    }
+
+    bool Uint(int u) {
+        vlog(cst_log.trace, "Uint({}), state: {}", u, _state);
+        switch (_state) {
+        case expectValue:
+            if ("version" == _key) {
+                if (u != static_cast<int>(topic_manifest_version::v1)) {
+                    throw std::runtime_error(
+                      "topic manifest version not supported");
+                }
+            } else if ("partition_count" == _key) {
+                _topic_config.partition_count = u;
+            } else if ("replication_factor" == _key) {
+                _topic_config.replication_factor = u;
+            } else if ("revision_id" == _key) {
+                _revision_id = u;
+            } else if ("segment_size" == _key) {
+                _topic_config.properties.segment_size = u;
+            } else if ("retention_bytes" == _key) {
+                _topic_config.properties.retention_bytes = tristate<size_t>(u);
+            } else if ("retention_duration" == _key) {
+                _topic_config.properties.retention_duration
+                  = tristate<std::chrono::milliseconds>(
+                    std::chrono::milliseconds(u));
+            } else {
+                return false;
+            }
+            if (_required_fields.contains(_key)) {
+                _seen_required_fields.insert(_key);
+            }
+            _state = all_required_complete() ? expectKeyOrObjectEnd : expectKey;
+            return true;
+        default:
+            return false;
+        }
+    }
+
+    bool EndObject(rapidjson::SizeType size) {
+        vlog(cst_log.trace, "EndObject({}), state_: {}", size, _state);
+        return _state == expectKeyOrObjectEnd;
+    }
+
+    bool Null() {
+        vlog(cst_log.trace, "Null(), state_: {}", _state);
+        _state = all_required_complete() ? expectKeyOrObjectEnd : expectKey;
+        return true;
+    }
+
+    bool Default() {
+        vlog(cst_log.trace, "Default(), state_: {}", _state);
+        return false;
+    }
+
+    bool all_required_complete() {
+        vlog(cst_log.trace, "headerIsComplete(), state_: {}", _state);
+        return _seen_required_fields.size() == _required_fields.size();
```

Comment:
We can store parsed field values as optionals, check in this function that all necessary fields are present and construct `topic_configuration` at the very end. This will allow us (for a price of slightly more tedious code) to get rid of default constructors for `topic_configuration` and `topic_namespace` and of std::set here.

Reply:
Fixed